src/NuGet.Core/NuGet.Configuration/Settings/SettingsLoadingContext.cs

```diff
@@ -30,46 +52,57 @@ internal SettingsFile GetOrCreateSettingsFile(string filePath, bool isMachineWid
 
             if (filePath == null)
             {
-                throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, Resources.Argument_Cannot_Be_Null_Or_Empty, nameof(filePath)));
+                throw new ArgumentNullException(nameof(filePath));
             }
 
-            _semaphore.Wait();
-            Lazy<SettingsFile> settingsFile = null;
-
-            try
+            lock (_lock)
             {
-                for (int i = 0; i < _settingsFiles.Count; i++)
+                if (!_fileCache.TryGetValue(filePath, out (DateTime LastWriteTime, SettingsFile SettingsFile) item))
                 {
-                    if (PathUtility.GetStringComparerBasedOnOS().Equals(_settingsFiles[i].Value.ConfigFilePath, filePath))
-                    {
-                        return _settingsFiles[i].Value;
-                    }
+                    SettingsFile settingsFile = LoadSettingsFile(new FileInfo(filePath), isMachineWide, isReadOnly);
+
+                    _fileCache.Add(filePath, (File.GetLastWriteTime(filePath), settingsFile));
+
+                    return settingsFile;
+                }
+
+                DateTime lastWriteTime = File.GetLastWriteTime(filePath);
```

Comment:
This doesn't handle FileNotFoundException, in case the config was deleted. Is that being handled somewhere else?